### PR TITLE
Added project.search method

### PR DESCRIPTION
**Pull request by artms**

### CODE REVIEW

File: projects_test.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
+package gonduit
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/uber/gonduit/core"
+	//	"github.com/uber/gonduit/entities"
```
Comment (artms) [inline projects_test.go:10]:
```suggestion
```

File: requests/project.go [hunk 2]
```diff
@@ -16,3 +19,54 @@ type ProjectQueryRequest struct {
 	Offset  uint64                  `json:"offset"`
 	Request
 }
+
+// DifferentialRevisionSearchRequest represents a request to
+// differential.revision.search API method.
```
Comment (artms) [inline requests/project.go:24]:
```suggestion
// ProjectSearchRequest represents a request to
// project.search API method.
```

File: responses/project.go [hunk 3]
```diff
@@ -1,10 +1,72 @@
...
 
 // ProjectQueryResponse represents a response from calling project.query.
 type ProjectQueryResponse struct {
 	Data    map[string]entities.Project `json:"data"`
 	SlugMap map[string]string           `json:"sligMap"`
 	Cursor  entities.Cursor             `json:"cursor"`
 }
+
+// ProjectSearchResponse contains fields that are in server
+// response to differential.diff.search.
```
Comment (artms) [inline responses/project.go:16]:
```suggestion
// response to project.search.
```

File: responses/project.go [hunk 4]
```diff
@@ -1,10 +1,72 @@
...
+	Name         string             `json:"name"`
+	Slug         string             `json:"slug"`
+	Description  string             `json:"description"`
+	Subtype      string             `json:"subtype"`
+	Milestone    *int               `json:"milestone"`
+	Depth        int                `json:"depth"`
+	Parent       *ProjectParent     `json:"parent"`
+	Icon         ProjectIcon        `json:"icon"`
+	Color        ProjectColor       `json:"color"`
+	SpacePHID    *string             `json:"spacePHID"`
```
Comment (artms) [inline responses/project.go:46]:
```suggestion
	SpacePHID    *string            `json:"spacePHID"`
```

File: projects.go [hunk 5]
```diff
@@ -5,13 +5,31 @@ import (
 	"github.com/uber/gonduit/responses"
 )
 
+const ProjectQueryMethod = "project.query"
```
Comment (Pawka) [inline projects.go:8]: missing comment for public variable.

File: projects_test.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}`
+
+func TestProjectSearch(t *testing.T) {
+	s := server.New()
+	defer s.Close()
+	s.RegisterCapabilities()
+	response := server.ResponseFromJSON(projectSearchResponseJSON)
+	s.RegisterMethod(ProjectSearchMethod, http.StatusOK, response)
+	number := 26
```
Comment (Pawka) [inline projects_test.go:100]: rename to "milestoneID" or something similar. The "number" is too vague.

File: requests/project.go [hunk 7]
```diff
@@ -16,3 +19,54 @@ type ProjectQueryRequest struct {
...
+	Attachments *ProjectSearchAttachments `json:"attachments,omitempty"`
+
+	*entities.Cursor
+	Request
+}
+
+// ProjectSearchAttachments contains fields that specify what
+// additional data should be returned with search results.
+type ProjectSearchAttachments struct {
+	// Reviewers requests to get the reviewers for each revision.
```
Comment (Pawka) [inline requests/project.go:43]: Misleading comment.

File: requests/project.go [hunk 8]
```diff
@@ -16,3 +19,54 @@ type ProjectQueryRequest struct {
...
+	Members bool `json:"members,omitempty"`
+	// Subscribers if true instructs server to return subscribers list for each task.
```
Comment (Pawka) [inline requests/project.go]: Misleading comment.

File: requests/project.go [hunk 9]
```diff
@@ -16,3 +19,54 @@ type ProjectQueryRequest struct {
...
+	Watchers bool `json:"watchers,omitempty"`
+	// Projects requests to get information about projects.
```
Comment (Pawka) [inline requests/project.go]: Misleading comment.

File: projects.go (hunk 5 above)
Comment (artms) [inline projects.go:8]:
```suggestion
// ProjectQueryMethod is method name on Phabricator API.
const ProjectQueryMethod = "project.query"
```

File: requests/project.go [hunk 10]
```diff
@@ -16,3 +19,54 @@ type ProjectQueryRequest struct {
...
+type ProjectSearchConstraints struct {
+	IDs   []int    `json:"ids,omitempty"`
+	PHIDs []string `json:"phids,omitempty"`
+	Name  string   `json:"name,omitempty"`
+	Slugs []string `json:"slugs,omitempty"`
+	// use phids
+	Members []string `json:"members,omitempty"`
+	// use phids
+	Watchers    []string `json:"watchers,omitempty"`
+	isMilestone bool     `json:"isMilestone,omitempty"`
```
Comment (Pawka) [inline requests/project.go:61]: Make this and other private variables public by starting with uppercase letter.

File: responses/project.go [hunk 11]
```diff
@@ -1,10 +1,72 @@
...
+}
+
+// ProjectSearchResponseItemFields is a collection of object
+// fields.
+type ProjectSearchResponseItemFields struct {
+	Name         string             `json:"name"`
+	Slug         string             `json:"slug"`
+	Description  string             `json:"description"`
+	Subtype      string             `json:"subtype"`
+	Milestone    *int               `json:"milestone"`
```
Comment (Pawka) [inline responses/project.go:41]: why this is a pointer?

File: responses/project.go [hunk 12]
```diff
@@ -1,10 +1,72 @@
...
+	Depth        int                `json:"depth"`
+	Parent       *ProjectParent     `json:"parent"`
+	Icon         ProjectIcon        `json:"icon"`
+	Color        ProjectColor       `json:"color"`
+	SpacePHID    *string            `json:"spacePHID"`
```
Comment (Pawka) [inline responses/project.go:46]: why this is a pointer?

File: responses/project.go (hunk 11 above)
Comment (artms) [reply responses/project.go:41]: Milestone can be `null`

File: responses/project.go (hunk 12 above)
Comment (artms) [reply responses/project.go:46]: SpacePHID can be null

Comment (artms) [inline responses/project.go:46]:
```suggestion
	SpacePHID     string            `json:"spacePHID"`
```

**Review by Pawka — APPROVED** [review]

**Pull request merged**